README
-*- restructuredtext -*-

Django HTTP Proxy
=================

Django HTTP Proxy provides simple HTTP proxy functionality for the Django web
development framework.

Motivation
----------

One possible use for this application (actually, the reason it was developed)
is to allow for easy development of Ajax applications against a live server
environment. 

The built-in `Django development server <http://docs.djangoproject.com/en/dev/ref/django-admin/#runserver>`_
already makes Django a wonderful tool for front-end developers needing a quick
way to set up a HTTP server. Now, with
`Django HTTP Proxy <http://bitbucket.org/yvandermeer/django-http-proxy/>`_ you
can seamlessly integrate your front-end work with a live back-end. No more
ugly PHP proxy scripts, and no more "standalone" vs "integrated" switches in a
client-side Communication Controller (unless you really want to of course!).

For instance, you might be developing an Ajax application which end up running
in an JSP environment. Instead of having to deploy your application on the JSP
server, you can already test against against the live server environment while
you develop. All you need is the Django framework, run `manage.py` and you are
good to go.

Installation
------------

Largely based on `a blog post by Will Larson <http://lethain.com/entry/2008/sep/30/suffer-less-by-using-django-dev-server-as-a-proxy/>`_,
Django HTTP Proxy is maintained as open-source software on Bitbucket:
http://bitbucket.org/yvandermeer/django-http-proxy/ 

It is also available on the Python Package Index:

http://pypi.python.org/pypi/django-http-proxy

As such, you can easily install the latest release using
`easy_install <http://pypi.python.org/pypi/setuptools>`_::

	$ easy_install django-http-proxy

or using `pip <http://pypi.python.org/pypi/pip>`_::

	$ pip install django-http-proxy

Usage
-----

For further documentation on installing Django HTTP Proxy and using it, see
the INSTALL file.
  
Contributing
------------

If you have any contributions, feel free to fork Django HTTP Proxy on
Bitbucket: 

http://bitbucket.org/yvandermeer/django-http-proxy/fork/



docs/changes.rst
Changes
-------

0.4.3 (2016-02-10)
~~~~~~~~~~~~~~~~~~

* Merged `pull request <https://github.com/yvandermeer/django-http-proxy/pull/20>`_ from `Ihor Kucher <https://github.com/crash843>`_ (thanks!): add support of post requests & update headers setting

0.4.2 (2016-02-10)
~~~~~~~~~~~~~~~~~~

* Merged `pull request <https://github.com/yvandermeer/django-http-proxy/pull/21>`_ from `Don Naegely <https://github.com/naegelyd>`_ (thanks!): Generate missing migrations

0.4.1
~~~~~

* Python 3 compatibility (requires Django >= 1.6)
* Fix duplicated forward slashes in urls

0.4
~~~

* Migration from Bitbucket (Mercurial) to Github
* Refactored main view using Django class-based views (see :class:`~httpproxy.views.HttpProxy`)
* Removed basic authentication support (`PROXY_USER` and `PROXY_PASSWORD`); may be added back later on.
* Finally merged back Django 1.6 fixes by `Petr Dlouhý <https://bitbucket.org/pdlouhy>`_ (thanks!)
* Merged pull request from `Garrett Seward <https://github.com/spectralsun>`_ (thanks!)
* Added Django 1.7 compatibility
* Added database migrations (Django 1.7 and higher only)
* Updated and improvement the documentation (including :doc:`API documentation <api/index>`)
* Added an ``example`` project for reference
* Using urllib2 instead of httplib2
* Using setuptools instead of distutils
* Using versioneer2 for package versioning
* Removed some unused imports and did some further code cleanup

0.3.2
~~~~~

* Limited display of request querystring in admin screen to 50 characters

0.3.1
~~~~~

* Fixed 250 character limitation for querystring in Recorded Request 
  (`issue #2 <http://bitbucket.org/yvandermeer/django-http-proxy/issue/2/>`_)
* Added new Request Parameter model; requires ``./manage.py reset httpproxy && ./manage.py syncdb``

0.3
~~~

* Fixed Python 2.5 support by removing use of ``__package__``
* Implemented request path "normalization", fixing record and playback if the
  proxy is URL-configured anywhere other than directly in the root.
* Added experimental ``PROXY_REWRITE_RESPONSES`` settings to fix paths to
  resources (images, javascript, etc) on the same domain if ``httproxy`` is
  not configured at the root.

0.2.2
~~~~~

* Removed print statement I accidentally left behind.

0.2.1
~~~~~

* Fixed `issue #1 <http://bitbucket.org/yvandermeer/django-http-proxy/issue/1/>`_;
  Unsupported content types are now silently ignored.
* Added ``PROXY_IGNORE_UNSUPPORTED`` setting to control the behavior for
  handling unsupported responses.

0.2
~~~

* Added recording and playback functionality
* Improved handling of ``httpproxy``-specific settings
* Started using Sphinx for documentation

0.1
~~~

* Initial release
* Basic HTTP proxy functionality based on `a blog post by Will Larson <http://lethain.com/entry/2008/sep/30/suffer-less-by-using-django-dev-server-as-a-proxy/>`_


httpproxy/models.py
from django.db import models
from django.utils.six.moves import urllib
from django.utils.translation import ugettext as _


class Request(models.Model):
    """
    An HTTP request recorded in the database.

    Used by the :class:`~httpproxy.recorder.ProxyRecorder` to record all
    identifying aspects of an HTTP request for matching later on when playing
    back the response.

    Request parameters are recorded separately, see
    :class:`~httpproxy.models.RequestParameter`.
    """
    method = models.CharField(_('method'), max_length=20)
    domain = models.CharField(_('domain'), max_length=100)
    port = models.PositiveSmallIntegerField(default=80)
    path = models.CharField(_('path'), max_length=250)
    date = models.DateTimeField(auto_now=True)
    querykey = models.CharField(_('query key'), max_length=255, editable=False)

    @property
    def querystring(self):
        """
        The URL-encoded set of request parameters.
        """
        return self.parameters.urlencode()

    def querystring_display(self):
        maxlength = 50
        if len(self.querystring) > maxlength:
            return '%s [...]' % self.querystring[:50]
        else:
            return self.querystring
    querystring_display.short_description = 'querystring'

    def __unicode__(self):
        output = u'%s %s:%d%s' % \
                (self.method, self.domain, self.port, self.path)
        if self.querystring:
            output += '?%s' % self.querystring
        return output[:50] # TODO add elipsed if truncating

    class Meta:
        verbose_name = _('request')
        verbose_name_plural = _('requests')
        unique_together = ('method', 'domain', 'port', 'path', 'querykey')
        get_latest_by = 'date'


class RequestParameterManager(models.Manager):

    def urlencode(self):
        output = []
        for param in self.values('name', 'value'):
            output.extend([urllib.parse.urlencode({param['name']: param['value']})])
        return '&'.join(output)


class RequestParameter(models.Model):
    """
    A single HTTP request parameter for a :class:`~httpproxy.models.Request`
    object.
    """
    REQUEST_TYPES = (
        ('G', 'GET'),
        ('P', 'POST'),
    )
    request = models.ForeignKey(Request, verbose_name=_('request'), related_name='parameters')
    type = models.CharField(max_length=1, choices=REQUEST_TYPES, default='G')
    order = models.PositiveSmallIntegerField(default=1)
    name = models.CharField(_('name'), max_length=100)
    value = models.CharField(_('value'), max_length=250, null=True, blank=True)
    objects = RequestParameterManager()

    def __unicode__(self):
        return u'%d %s=%s' % (self.pk, self.name, self.value)

    class Meta:
        ordering = ('order',)
        verbose_name = _('request parameter')
        verbose_name_plural = _('request parameters')


class Response(models.Model):
    """
    The response that was recorded in response to the corresponding
    :class:`~httpproxy.models.Request`.
    """
    request = models.OneToOneField(Request, verbose_name=_('request'))
    status = models.PositiveSmallIntegerField(default=200)
    content_type = models.CharField(_('content type'), max_length=200)
    content = models.TextField(_('content'))

    @property
    def request_domain(self):
        return self.request.domain

    @property
    def request_path(self):
        return self.request.path

    @property
    def request_querystring(self):
        return self.request.querystring

    def __unicode__(self):
        return u'Response to %s (%d)' % (self.request, self.status)

    class Meta:
        verbose_name = _('response')
        verbose_name_plural = _('responses')
